internal/daemon/controller/handlers/targets/target_service.go

```diff
@@ -234,87 +234,115 @@ func (s Service) ListTargets(ctx context.Context, req *pbs.ListTargetsRequest) (
 		return &pbs.ListTargetsResponse{}, nil
 	}
 
-	repo, err := s.repoFn(target.WithPermissions(userPerms))
+	pageSize := int(s.maxPageSize)
+	// Use the requested page size only if it is smaller than
+	// the configured max.
+	if req.GetPageSize() != 0 && uint(req.GetPageSize()) < s.maxPageSize {
+		pageSize = int(req.GetPageSize())
+	}
+	filter, err := handlers.NewFilter(ctx, req.GetFilter())
 	if err != nil {
 		return nil, err
 	}
-	filter, err := handlers.NewFilter(ctx, req.GetFilter())
+
+	repo, err := s.repoFn(target.WithPermissions(userPerms))
 	if err != nil {
 		return nil, err
 	}
-	listTargetsFn := func(prevPageLast target.Target, refreshToken *pbs.ListRefreshToken, limit int) ([]target.Target, error) {
-		opts := []target.Option{
-			target.WithLimit(limit),
+	filterItemFn := func(item target.Target) (bool, error) {
+		pbItem, err := toProto(ctx, item, newOutputOpts(ctx, item, authResults, authzScopes)...)
+		if err != nil {
+			return false, err
 		}
-		if prevPageLast == nil {
-			if refreshToken != nil {
-				opts = append(opts, target.WithStartPageAfterItem(
-					refreshToken.GetLastItemId(),
-					refreshToken.GetLastItemUpdatedTime().AsTime(),
-				))
-			}
-		} else {
-			opts = append(opts, target.WithStartPageAfterItem(prevPageLast.GetPublicId(), prevPageLast.GetUpdateTime().AsTime()))
+
+		filterable, err := subtypes.Filterable(pbItem)
+		if err != nil {
+			return false, err
 		}
-		return repo.ListTargets(ctx, opts...)
+		return filter.Match(filterable), nil
 	}
-	filterAndConvertFn := func(item target.Target) (*pb.Target, error) {
-		pr := perms.Resource{Id: item.GetPublicId(), ScopeId: item.GetProjectId(), Type: resource.Target}
-		outputFields := authResults.FetchOutputFields(pr, action.List).SelfOrDefaults(authResults.UserId)
 
-		outputOpts := make([]handlers.Option, 0, 3)
-		outputOpts = append(outputOpts, handlers.WithOutputFields(outputFields))
-
-		if outputFields.Has(globals.ScopeField) {
-			outputOpts = append(outputOpts, handlers.WithScope(authzScopes[item.GetProjectId()]))
-		}
-		if outputFields.Has(globals.AuthorizedActionsField) {
-			authorizedActions := authResults.FetchActionSetForId(ctx, item.GetPublicId(), IdActions, auth.WithResource(&pr)).Strings()
-			outputOpts = append(outputOpts, handlers.WithAuthorizedActions(authorizedActions))
+	var listResp *target.ListResponse
+	if req.GetRefreshToken() == "" {
+		listResp, err = target.List(ctx, repo, &authResults, pageSize, filterItemFn)
+		if err != nil {
+			return nil, err
 		}
-
-		pbItem, err := toProto(ctx, item, outputOpts...)
+	} else {
+		rt, err := parseRefreshToken(ctx, req.GetRefreshToken())
```

Comment:
The parsing of the refresh token from a string into a API level protobuf type remains the responsibility of the application layer.